### DISCUSSION TITLE
v2.0.0-alpha.3 -2023-04-13

### DISCUSSION THREAD

**Discussion by bryanculver** [nautobot/nautobot#3578] [General]
# 2.0 IPAM Overhaul Alpha Release

## Overview

This alpha introduces the Namespace concept to IPAM and makes many robust improvements to relationships between models:

- Namespace introduced to provide a uniqueness boundary for IPAddresses and Prefixes
- Prefixes have concrete relationships to their parents instead of derived
- IPAddresses have concrete relationships to their parent prefixes
- VRF is no longer used as a uniqueness constraint for IPAddresses and Prefixes

**⚠️ NOTE:** Because we are validating these new data models the data migrations have not been implemented. You will have to re-input your data largely via the UI to prove if these models will work for you. We will implement the migrations in an upcoming beta if the relationships between Namespace and other IPAM objects prove to be the right design. As such, do not expect to migrate future releases on these installs. The data migration steps won't be re-performed.

**For more detailed information, please view: https://github.com/nautobot/nautobot/issues/1032, https://github.com/nautobot/nautobot/issues/3337#issuecomment-1450855876**

❗ This has been recreated to fix a bug with the prerelease workflow.

---

### Changelog
* Add missing plugin links to docs page menu by @matt852 in https://github.com/nautobot/nautobot/pull/3447
* Removed  `block sidebar` from aggregate_list.html and objectchange_list.html. by @HanlinMiao in https://github.com/nautobot/nautobot/pull/3463
* Implement `Prefix.parent` concrete foreign key to self. by @jathanism in https://github.com/nautobot/nautobot/pull/3452
* IPAddress lookup fix by @glennmatthews in https://github.com/nautobot/nautobot/pull/3487
* Fix Docker Prerelease Tag Syntax by @bryanculver in https://github.com/nautobot/nautobot/pull/3475
* Fixes to Prefix views and functionality by @glennmatthews in https://github.com/nautobot/nautobot/pull/3496
* Update README.md by @lampwins in https://github.com/nautobot/nautobot/pull/3515
* Release 2.0.0-alpha.2 by @jathanism in https://github.com/nautobot/nautobot/pull/3511
* Introduced `IPAddress.parent` concrete foreign key to `Parent` by @jathanism in https://github.com/nautobot/nautobot/pull/3489
* VRF namespace enhancements by @glennmatthews in https://github.com/nautobot/nautobot/pull/3505
* Merge develop into next for docker tag fix by @gsnider2195 in https://github.com/nautobot/nautobot/pull/3522
* Bump redis from 4.3.4 to 4.5.4 by @dependabot in https://github.com/nautobot/nautobot/pull/3499
* Remove slug from models by @timizuoebideri1 in https://github.com/nautobot/nautobot/pull/3439
* Release/1.5.15 by @bryanculver in https://github.com/nautobot/nautobot/pull/3537
* Main to develop post v1.5.15 by @bryanculver in https://github.com/nautobot/nautobot/pull/3539
* Add note to ViewSet Attributes. by @scetron in https://github.com/nautobot/nautobot/pull/3478
* Remove nat_outside by @timizuoebideri1 in https://github.com/nautobot/nautobot/pull/3523
* Adds doc page for Circuit Maintenance by @jvanderaa in https://github.com/nautobot/nautobot/pull/3558
* Drop python 3.7 support for 2.0 by @glennmatthews in https://github.com/nautobot/nautobot/pull/3554
* Update README.md by @lampwins in https://github.com/nautobot/nautobot/pull/3564
* Use plugins `next-2.0` branch for `next` upstream tests by @snaselj in https://github.com/nautobot/nautobot/pull/3502
* Batch of things to improve working with relationships. by @bryanculver in https://github.com/nautobot/nautobot/pull/3550
* Fix #3525 - add missing dependency by @glennmatthews in https://github.com/nautobot/nautobot/pull/3563
* Release/1.5.16 by @bryanculver in https://github.com/nautobot/nautobot/pull/3565
* Main to develop post v1.5.16 by @bryanculver in https://github.com/nautobot/nautobot/pull/3569
* Defense against #3525 - add explicit dependency to develop as well by @glennmatthews in https://github.com/nautobot/nautobot/pull/3566
* Develop to next post 1.5.16 by @bryanculver in https://github.com/nautobot/nautobot/pull/3570
* [REVIEW ONLY] `next` into Namespaces Prototype by @bryanculver in https://github.com/nautobot/nautobot/pull/3572

**Full Changelog**: https://github.com/nautobot/nautobot/compare/v2.0.0-alpha.2...v2.0.0-alpha.3

<hr /><em>This discussion was created from the release <a href='https://github.com/nautobot/nautobot/releases/tag/v2.0.0-alpha.3'>v2.0.0-alpha.3 -2023-04-13</a>.</em>

**Comment by kingfetty** [comment]
Addresses still require CIDR notation but the CIDR notation is not needed as it does not appear to be taken into account in any of the calculation for parent/child relationships.
example: Prefix 192.168.1.0/24
Address[0]  192.168.1.1/24
Address[1]  192.168.1.2/32
Address[2]  192.168.1.3/16
Address[3]  192.168.1.4/27
All these were acceptable. When in reality if these were true CIDRs of these it would create issues.
Suggestion, pull the CIDR from the Prefix since there is a required parent anyway.  Or if the CIDR is smaller than the prefix, have the option of creating a new prefix along with the address?

**Reply by jathanism** [reply]
@kingfetty Thanks for the observations! We've had discussions about this and the reality is that the prefix length for an IP address does not matter. In actuality or even in practice it's really just shorthand for the netmask you're specifying when configuring that IP on a device--and exactly to your point--we could just inherit it from the parent `Prefix`.

Obviously at this time we have not yet removed it, but very well may do so.

One last thing, however, is that having the prefix length on the IP address as it is currently does not impact the CIDR notation or parenting algorithm whatsoever because all IPs are leaf nodes and cannot have children. While it might be confusing to see a `Prefix` with address `192.168.1.2/31` (which is valid, containing `.2` and `.3` as children) an `IPAddress` with the same address would merely represent `192.168.1.2` with a netmask of `255.255.255.254`, indicating that its parent is the `Prefix` of the same length. Because it is confusing however, is a good reason to change the semantics in the data design and user experience in Nautobot 2.0! :)
